internal/customermanagedkeys/key_vault_or_managed_hsm_key.go

```diff
@@ -0,0 +1,195 @@
+package customermanagedkeys
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/go-azure-sdk/sdk/environments"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/keyvault/parse"
+	hsmParse "github.com/hashicorp/terraform-provider-azurerm/internal/services/managedhsm/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+type VersionType int
+
+const (
+	VersionTypeAny VersionType = iota
+	VersionTypeVersioned
+	VersionTypeVersionless
+)
+
+type KeyVaultOrManagedHSMKey struct {
+	KeyVaultKeyId              *parse.NestedItemId
+	ManagedHSMKeyId            *hsmParse.ManagedHSMDataPlaneVersionedKeyId
+	ManagedHSMKeyVersionlessId *hsmParse.ManagedHSMDataPlaneVersionlessKeyId
+}
+
+func (k *KeyVaultOrManagedHSMKey) IsSet() bool {
+	return k != nil && (k.KeyVaultKeyId != nil || k.ManagedHSMKeyId != nil || k.ManagedHSMKeyVersionlessId != nil)
+}
+
+func (k *KeyVaultOrManagedHSMKey) ID() string {
+	if k == nil {
+		return ""
+	}
+
+	if k.KeyVaultKeyId != nil {
+		return k.KeyVaultKeyId.ID()
+	}
+
+	if k.ManagedHSMKeyId != nil {
+		return k.ManagedHSMKeyId.ID()
+	}
+
+	if k.ManagedHSMKeyVersionlessId != nil {
+		return k.ManagedHSMKeyVersionlessId.ID()
+	}
+
+	return ""
+}
+
+func (k *KeyVaultOrManagedHSMKey) KeyVaultKeyID() string {
+	if k != nil && k.KeyVaultKeyId != nil {
+		return k.KeyVaultKeyId.ID()
+	}
+	return ""
+}
+
+func (k *KeyVaultOrManagedHSMKey) ManagedHSMKeyID() string {
+	if k != nil && k.ManagedHSMKeyId != nil {
+		return k.ManagedHSMKeyId.ID()
+	}
+
+	if k != nil && k.ManagedHSMKeyVersionlessId != nil {
+		return k.ManagedHSMKeyVersionlessId.ID()
+	}
+
+	return ""
+}
+
+func (k *KeyVaultOrManagedHSMKey) BaseUri() string {
+	if k.KeyVaultKeyId != nil {
+		return k.KeyVaultKeyId.KeyVaultBaseUrl
+	}
+
+	if k.ManagedHSMKeyId != nil {
+		return k.ManagedHSMKeyId.BaseUri()
+	}
+
+	if k.ManagedHSMKeyVersionlessId != nil {
+		return k.ManagedHSMKeyVersionlessId.BaseUri()
+	}
+
+	return ""
+}
+
+func parseKeyvauleID(keyRaw string, requireVersion VersionType) (*parse.NestedItemId, error) {
```

Comment:
we're going to need the Key Vault Environment here in the future, for when the Key Vault nested items take the environment, so we should probably thread that through now, since this'll be used in a bunch of places?